### DISCUSSION TITLE
Can I plot nFrags on UMAP embedding?

### DISCUSSION THREAD

**Discussion by dmsalsgh97** [GreenleafLab/ArchR#743] [Questions / Documentation] [answered]
Hi, first thanks for developing and maintaining ArchR! 
ArchR  is the awesome reproducible workflow for scATAC!

Now, I want to see library depth(~  nFrags ) on current UMAP embedding, 
because my samples have very distinct library depth and I wonder to check my clusters were not affected by the distinct library depth.
In my think, required information already exists. (UMAP embedding, nFrags on cellColData)

But, when I tried the code below, it failes...

`p <- plotEmbedding(
    ArchRProj = projNX1_RmDublet, 
    colorBy = "cellColData",
    name = "nFrags", 
    embedding = "UMAP"
)
`
`
plotPDF(plotList = p, 
    name = "Plot-nFrags.pdf", 
    ArchRProj = projNX1_RmDublet, 
    addDOC = FALSE, width = 5, height = 5)`

When I run this code, the first-line runs without error, but the second-line outs error...

`
Plotting Other  
list()  
attr(,"class")  
[1] "waiver"  
Plotting Ggplot!  
no applicable method for 'ggplot_build' applied to an object of class "c('LayerInstance', 'Layer', 'ggproto', 'gg')"`

Can I get any help?

**Comment by rcorces** [comment, marked as answer]
Thanks for using ArchR!
You've mis-used the parameters in `plotPDF()`. `plotList` is only to be used when providing a list of plots and you do not have a list of plots - hence the error. https://www.archrproject.com/reference/plotPDF.html

```
proj <- getTestProject()

p1 <- plotEmbedding(
  ArchRProj = proj, 
  colorBy = "cellColData", 
  name = "nFrags", 
  embedding = "UMAP",
  colorLimits = c(0,5)
)

plotPDF(p1, name = "nfrags.pdf", ArchRProj = proj, addDOC=FALSE, width = 5, height = 5)
```

**Reply by dmsalsgh97** [reply]
Oops. Thanks for quick reply!
I've searched this error message, but I couldn't find the cause. .

**Comment by dmsalsgh97** [comment]
Hi, Ryan, I have additional questions about dealing with depth(~ nFrags ) on the ArchR.

In Signac vignettes, they are visualizing Correlation between depth/reduced dim components and removing strong correlative components for downstream analysis.

![](https://satijalab.org/signac/articles/pbmc_vignette_files/figure-html/unnamed-chunk-12-1.png)

I found that ArchR saves its LSI pieces of information on ~/IterativeLSI/Save-LSI-Iteration-1.rds, and the LSI object has the corToDepth feature.

```> LSI$LSI$corToDepth
$scaled
        LSI1         LSI2         LSI3         LSI4         LSI5         LSI6 
0.1128479858 0.2289701114 0.0111613698 0.1153006097 0.0140032424 0.0446032240 
        LSI7         LSI8         LSI9        LSI10        LSI11        LSI12 
0.0992681142 0.1137278129 0.0114642776 0.0982701492 0.0016834675 0.0631183438 
       LSI13        LSI14        LSI15        LSI16        LSI17        LSI18 
0.0171417867 0.0269677572 0.0471520192 0.0419728044 0.0314250424 0.0536640915 
       LSI19        LSI20        LSI21        LSI22        LSI23        LSI24 
0.0214370762 0.0418425570 0.0064629541 0.0227947271 0.0383506661 0.0516500763 
       LSI25        LSI26        LSI27        LSI28        LSI29        LSI30 
0.0572827479 0.0366720083 0.0239598162 0.0118163372 0.0116859520 0.0009461114 

$none
        LSI1         LSI2         LSI3         LSI4         LSI5         LSI6 
0.8899458037 0.2714577888 0.1663154065 0.3053934210 0.0041466663 0.0772972687 
        LSI7         LSI8         LSI9        LSI10        LSI11        LSI12 
0.1686316215 0.0802010044 0.0323139501 0.0169087629 0.0564525639 0.0352069392 
       LSI13        LSI14        LSI15        LSI16        LSI17        LSI18 
0.0555790464 0.0056453964 0.0173366740 0.0597627408 0.0308390907 0.0106807972 
       LSI19        LSI20        LSI21        LSI22        LSI23        LSI24 
0.0132201590 0.0758390729 0.0688394275 0.0490308403 0.0518457268 0.0645299437 
       LSI25        LSI26        LSI27        LSI28        LSI29        LSI30 
0.0105281067 0.0605595129 0.0418302591 0.0009725525 0.0867666523 0.0038796550 
```
Q1. In my think, this works as selecting feature of 2 iterations of the LSI calculations, and in my case, $scaled is used because it has lower and even corToDepth. is it right?

Q2. In Signac vignettes, there are both positive and negative correlations, but in my LSI calculations, there is only positive correlations... is it possible?

So, I can expect that my LSI calculation has minimal effects on depth.
But, When I plot UMAP embedding with colData$Sample and colData$nFrags, I can see depth is significantly affecting UMAP distribution and clustering...
![image](https://user-images.githubusercontent.com/42495757/117752696-361f1080-b252-11eb-8cb1-0aa44a58c3ff.png)

Q3. In that case, should I run cellranger-aggr for hard depth normalization( remove reads ) for meaningful analysis?



Below is my scripts.
```
projNX1_RmDublet <- addIterativeLSI(  
    ArchRProj = projNX1_RmDublet,  
    useMatrix = "TileMatrix",   
    name = "IterativeLSI",   
    iterations = 2,   
    clusterParams = list( #See Seurat::FindClusters  
        resolution = c(0.2),   
        sampleCells = 10000,   
        n.start = 10  
    ), 
    varFeatures = 25000, 
    dimsToUse = 1:30
)

projNX1_RmDublet <- addClusters(
    input = projNX1_RmDublet,
    reducedDims = "IterativeLSI",
    method = "Seurat",
    name = "Clusters",
    resolution = 0.8
) 

projNX1_RmDublet <- addUMAP(
    ArchRProj = projNX1_RmDublet, 
    reducedDims = "IterativeLSI", 
    name = "UMAP", 
    nNeighbors = 30, 
    minDist = 0.5, 
    metric = "cosine"
)
```

I'm so sorry for the detailed questions about my research...

**Reply by rcorces** [reply]
> Q1. In my think, this works as selecting feature of 2 iterations of the LSI calculations, and in my case, $scaled is used because it has lower and even corToDepth. is it right?

I'm not sure I understand the question but I'd encourage you to read through the function/parameter definitions. In particular:


<html>
<body>


scaleDims | A boolean that indicates whether to z-score the reduced dimensions for each cell. This is useful forminimizing the contribution of strong biases (dominating early PCs) and lowly abundant populations. However, this may lead to stronger sample-specific biases since it is over-weighting latent PCs. If set to NULL this will scale the dimensions based on the value of scaleDims when the reducedDims were originally created during dimensionality reduction. This idea was introduced by Timothy Stuart.
-- | --



</body>
</html>

<html>
<body>


corCutOff | A numeric cutoff for the correlation of each dimension to the sequencing depth. If the dimension has a correlation to sequencing depth that is greater than the corCutOff, it will be excluded from analysis.
-- | --



</body>
</html>

> Q2. In Signac vignettes, there are both positive and negative correlations, but in my LSI calculations, there is only positive correlations... is it possible?

The `corToDepth` is taken as the absolute value of the correlation so negative correlations are shown as positive:
https://github.com/GreenleafLab/ArchR/blob/968e4421ce7187a8ac7ea1cf6077412126876d5f/R/IterativeLSI.R#L565

> Q3. In that case, should I run cellranger-aggr for hard depth normalization( remove reads ) for meaningful analysis?

No - ArchR is not meant to be run with cellranger-aggr. Its important to note that ArchR automatically handles dimensions correlated to depth above the threshold that you set with `corCutoff`

**Reply by dmsalsgh97** [reply]
Thanks for the reply!
I'll study more about scATAC-seq analysis. :)
